crates/fuel-core/src/service/adapters.rs

```diff
@@ -85,6 +86,155 @@ impl StaticGasPrice {
     }
 }
 
+#[cfg(test)]
+mod arc_gas_price_estimate_tests {
+    #![allow(non_snake_case)]
+
+    use super::*;
+    use proptest::proptest;
+
+    async fn _worst_case__correctly_calculates_value(
+        gas_price: u64,
+        starting_height: u32,
+        block_horizon: u32,
+        percentage: u16,
+    ) {
+        // given
+        let subject = ArcGasPriceEstimate::new(starting_height, gas_price, percentage);
+
+        // when
+        let target_height = starting_height.saturating_add(block_horizon);
+        let actual = subject
+            .worst_case_gas_price(target_height.into())
+            .await
+            .unwrap();
+
+        // then
+        let mut expected = gas_price;
+
+        for _ in 0..block_horizon {
+            let change_amount = expected
+                .saturating_mul(percentage as u64)
+                .saturating_div(100);
+            expected = expected.saturating_add(change_amount);
+        }
+
+        assert!(actual >= expected);
+    }
+
+    proptest! {
+        #[test]
+        fn worst_case_gas_price__correctly_calculates_value(
+            gas_price: u64,
+            starting_height: u32,
+            block_horizon in 0..10_000u32,
+            percentage: u16,
+        ) {
+            let rt = tokio::runtime::Runtime::new().unwrap();
+            rt.block_on(_worst_case__correctly_calculates_value(
+                gas_price,
+                starting_height,
+                block_horizon,
+                percentage,
+            ));
+        }
+    }
+
+    proptest! {
+        #[test]
+        fn worst_case_gas_price__never_overflows(
+            gas_price: u64,
+            starting_height: u32,
+            block_horizon in 0..10_000u32,
+            percentage: u16
+        ) {
+            let rt = tokio::runtime::Runtime::new().unwrap();
+
+            // given
+            let subject = ArcGasPriceEstimate::new(starting_height, gas_price, percentage);
+
+            // when
+            let target_height = starting_height.saturating_add(block_horizon);
+
+            let _ = rt.block_on(subject.worst_case_gas_price(target_height.into()));
+
+            // then
+            // doesn't panic with an overflow
+        }
+    }
+}
+
+#[allow(dead_code)]
+pub struct ArcGasPriceEstimate<Height, GasPrice> {
```

Comment:
would it be possible to have a bit of documentation here? 
In particular, I'd like to understand what the "best_height" is here